File: Archive_LJ_grand_canonical/ggcmclj.f90
!**********************************************************************!
!
! File: ggcmclj.f90
!
! Create random ("gas") initial configuration for Grand Canonical (TVmu)
! Monte Carlo of Lennard-Jonesium
!
! 04-Feb-2004 (MN)
! 18-Apr-2012
!
!**********************************************************************!

program ggcmclj

!**********************************************************************!

  implicit none

  integer,parameter::iout=2

  character(len=80)::fname
  integer::i,n,nran,nt,ntjob,ntprint,ntskip
  integer,dimension(:),allocatable::iran
  real::c,disp,dr,pcre,ran,t,v,zz
  real,dimension(:),allocatable::x,y,z

! User input

  write(unit=*,fmt="(a)",advance="no") "              t="
  read(unit=*,fmt=*) t
  write(unit=*,fmt="(a)",advance="no") "              v="
  read(unit=*,fmt=*) v
  write(unit=*,fmt="(a)",advance="no") "              z="
  read(unit=*,fmt=*) zz
  write(unit=*,fmt="(a)",advance="no") "              n="
  read(unit=*,fmt=*) n

  do
    write(unit=*,fmt="(a)",advance="no") "     p(cre/del)="
    read(unit=*,fmt=*) pcre
    if(pcre<0.5) then
      exit
    end if
  end do

  write(unit=*,fmt="(a)",advance="no") "           disp="
  read(unit=*,fmt=*) disp
  write(unit=*,fmt="(a)",advance="no") "             dr="
  read(unit=*,fmt=*) dr
  write(unit=*,fmt="(a)",advance="no") "         ntskip="
  read(unit=*,fmt=*) ntskip
  write(unit=*,fmt="(a)",advance="no") " ntprint/ntskip="
  read(unit=*,fmt=*) ntprint
  write(unit=*,fmt="(a)",advance="no") "   ntjob/ntskip="
  read(unit=*,fmt=*) ntjob

  write(unit=*,fmt="(a)",advance="no") "          fname=[gcmclj_in.dat] "
  read(unit=*,fmt="(a)") fname
  if(fname=="") then
    fname="gcmclj_in.dat"
  end if

! RNG characteristics

  call random_seed(size=nran)

! Allocate arrays

  allocate(iran(nran),x(n),y(n),z(n))

! Box length

  c=v**(1.0/3.0)
  
! Random particle positions

  do i=1,n
    call random_number(ran)
    x(i)=(ran-0.5)*c
    call random_number(ran)
    y(i)=(ran-0.5)*c
    call random_number(ran)
    z(i)=(ran-0.5)*c
  end do

! RNG seed

  call random_seed(get=iran)

! Write startup file

  nt=0

  open(unit=iout,file=fname,status="replace",action="write", &
       form="unformatted")

  write(unit=iout) n,nt,ntjob,ntprint,ntskip,iran
  write(unit=iout) disp,dr,pcre,t,v,zz
  write(unit=iout) x,y,z

  close(unit=iout)

! Deallocate arrays

  deallocate(iran,x,y,z)

end program ggcmclj

!**********************************************************************!

File: Archive_LJ_grand_canonical/ggcmclj.c
/**********************************************************************
 *
 * File: ggcmclj.c
 *
 * Create random ("gas") initial configuration for Grand Canonical
 * (TVmu) Monte Carlo of Lennard-Jonesium
 *
 * 08-Apr-2010 (MN)
 * 18-Apr-2012
 *
 **********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "getval.h"

#define BSIZE 80


/**********************************************************************/

int main() {

/**********************************************************************/

  FILE *fpo;
  char fname[BSIZE];
  unsigned short dummy[3];
  unsigned short *seed;
  int n,nt,ntjob,ntprint,ntskip;
  int i;
  double pcre,disp,dr,t,v,zz;
  double c;
  double *x,*y,*z;

  /* User input */

  printf("              t=");
  getval_d(&t);
  printf("              v=");
  getval_d(&v);
  printf("              z=");
  getval_d(&zz);
  printf("              n=");
  getval_i(&n);

  for(;;) {
    printf("     p(cre/del)=");
    getval_d(&pcre);
    if(pcre<0.5) {
      break;
    }
  }

  printf("           disp=");
  getval_d(&disp);
  printf("             dr=");
  getval_d(&dr);
  printf("         ntskip=");
  getval_i(&ntskip);
  printf(" ntprint/ntskip=");
  getval_i(&ntprint);
  printf("   ntjob/ntskip=");
  getval_i(&ntjob);

  strcpy(fname,"gcmclj_in.dat");
  printf("          fname=[gcmclj_in.dat] ");
  getval_s(fname);

  /* Allocate arrays */

  x=(double*)malloc(n*sizeof(double));
  y=(double*)malloc(n*sizeof(double));
  z=(double*)malloc(n*sizeof(double));

  /* Box length */

  c=cbrt(v);

  /* Random particle positions */

  for(i=0;i<=n-1;i++) {
    x[i]=(drand48()-0.5)*c;
    y[i]=(drand48()-0.5)*c;
    z[i]=(drand48()-0.5)*c;
  }

  /* RNG seed */

  seed=seed48(dummy);

  /* Write startup file */

  nt=0;

  fpo=fopen(fname,"w");

  fwrite(&n,sizeof(int),1,fpo);
  fwrite(&nt,sizeof(int),1,fpo);
  fwrite(&ntjob,sizeof(int),1,fpo);
  fwrite(&ntprint,sizeof(int),1,fpo);
  fwrite(&ntskip,sizeof(int),1,fpo);
  fwrite(seed,sizeof(unsigned short),3,fpo);

  fwrite(&disp,sizeof(double),1,fpo);
  fwrite(&dr,sizeof(double),1,fpo);
  fwrite(&pcre,sizeof(double),1,fpo);
  fwrite(&t,sizeof(double),1,fpo);
  fwrite(&v,sizeof(double),1,fpo);
  fwrite(&zz,sizeof(double),1,fpo);

  fwrite(x,sizeof(double),n,fpo);
  fwrite(y,sizeof(double),n,fpo);
  fwrite(z,sizeof(double),n,fpo);

  fclose(fpo);

  /* Deallocate arrays */

  free(x);
  free(y);
  free(z);

  return 0;

}

/**********************************************************************/
